File: src/main/scala/com/xmlcalabash/model/xml/XMLViewportComposer.scala
package com.xmlcalabash.model.xml

import com.jafpl.messages.{Message, Metadata}
import com.jafpl.steps.{ViewportComposer, ViewportItem}
import com.xmlcalabash.config.XMLCalabashConfig
import com.xmlcalabash.exceptions.XProcException
import com.xmlcalabash.messages.{XdmNodeItemMessage, XdmValueItemMessage}
import com.xmlcalabash.model.util.{SaxonTreeBuilder, ValueParser, XProcConstants}
import com.xmlcalabash.runtime.{ProcessMatch, ProcessMatchingNodes, StaticContext, XProcMetadata, XProcVtExpression, XProcXPathExpression}
import com.xmlcalabash.util.{TypeUtils, VoidLocation}
import net.sf.saxon.om.{AttributeMap, SingletonAttributeMap}
import net.sf.saxon.s9api.{QName, XdmAtomicValue, XdmNode, XdmValue}

import scala.collection.mutable
import scala.collection.mutable.ListBuffer

class XMLViewportComposer(config: XMLCalabashConfig, context: StaticContext, patternString: String) extends ViewportComposer {
  private val cx_viewport = new QName("cx", XProcConstants.ns_cx,"viewport")
  private val _index = new QName("index")
  private var matcher: ProcessMatch = _
  private var decomposed: XdmNode = _
  private var metadata: XProcMetadata = _
  private var items = ListBuffer.empty[XMLViewportItem]
  private var itemIndex = 0
  private var dynBindings = Map.empty[String, Message]

  override def runtimeBindings(bindings: Map[String, Message]): Unit = {
    dynBindings = bindings
  }

  override def decompose(message: Message): List[ViewportItem] = {
    var source: XdmNode = null

    message match {
      case msg: XdmNodeItemMessage =>
        source = msg.item
        metadata = msg.metadata
      case msg: XdmValueItemMessage =>
        throw XProcException.xdBadViewportInput(msg.metadata.contentType, context.location)
      case _ =>
        throw XProcException.xiThisCantHappen("Viewport message without metadata?", context.location)
    }

    val bindings = mutable.HashMap.empty[String,Message] ++ context.statics
    for ((name, message) <- dynBindings) {
      bindings.put(name,message)
    }

    /*
    val expr = new XProcXPathExpression(context, patternString)
    var msg = config.expressionEvaluator.singletonValue(expr, List(), bindings.toMap, None)
    // Ok, now we have a string value
    pattern = msg.item.getUnderlyingValue.getStringValue
    */

    matcher = new ProcessMatch(config, new Decomposer(), context, bindings.toMap)
    matcher.process(source, patternString)
    decomposed = matcher.result
    items.toList
  }

  override def recompose(): Message = {
    matcher = new ProcessMatch(config, new Recomposer(), context)
    matcher.process(decomposed, "*") // FIXME: match only cx:viewport
    val recomposed = matcher.result
    new XdmNodeItemMessage(recomposed, metadata, context)
  }

  private class Decomposer() extends ProcessMatchingNodes {
    private def insertMarker(): Unit = {
      matcher.addStartElement(cx_viewport, SingletonAttributeMap.of(TypeUtils.attributeInfo(_index, itemIndex.toString)))
      itemIndex += 1
    }

    override def startDocument(node: XdmNode): Boolean = {
      items += new XMLViewportItem(node, metadata)
      insertMarker()
      matcher.addEndElement()
      false
    }

    override def startElement(node: XdmNode, attributes: AttributeMap): Boolean = {
      items += new XMLViewportItem(node, metadata.withBaseURI(Option(node.getBaseURI)))
      insertMarker()
      false
    }

    override def endElement(node: XdmNode): Unit = {
      matcher.addEndElement()
    }

    override def endDocument(node: XdmNode): Unit = {
      matcher.endDocument()
    }

    override def attributes(node: XdmNode, matchingAttributes: AttributeMap, nonMatchingAttributes: AttributeMap): Option[AttributeMap] = {
      throw XProcException.xdViewportOnAttribute(patternString, context.location)
    }

    override def text(node: XdmNode): Unit = {
      items += new XMLViewportItem(node, metadata.withBaseURI(Option(node.getBaseURI)))
      insertMarker()
      matcher.addEndElement()
    }

    override def comment(node: XdmNode): Unit = {
      items += new XMLViewportItem(node, metadata.withBaseURI(Option(node.getBaseURI)))
      insertMarker()
      matcher.addEndElement()
    }

    override def pi(node: XdmNode): Unit = {
      items += new XMLViewportItem(node, metadata.withBaseURI(Option(node.getBaseURI)))
      insertMarker()
      matcher.addEndElement()
    }
  }

  private class Recomposer() extends ProcessMatchingNodes {
    private def processMarker(marker: XdmNode): Unit = {
      val index = marker.getAttributeValue(_index).toInt
      matcher.addSubtree(items(index).getReplacement)
    }

    override def startDocument(node: XdmNode): Boolean = {
      matcher.startDocument(node.getBaseURI)
      true
    }

    override def startElement(node: XdmNode, attributes: AttributeMap): Boolean = {
      if (node.getNodeName == cx_viewport) {
        processMarker(node)
        false
      } else {
        matcher.addStartElement(node)
        true
      }
    }

    override def endElement(node: XdmNode): Unit = {
      if (node.getNodeName != cx_viewport) {
        matcher.addEndElement()
      }
    }

    override def endDocument(node: XdmNode): Unit = {
      matcher.endDocument()
    }

    override def attributes(node: XdmNode, matchingAttributes: AttributeMap, nonMatchingAttributes: AttributeMap): Option[AttributeMap] = {
      throw XProcException.xcInvalidSelection(patternString, "attribute", None)
    }

    override def text(node: XdmNode): Unit = {
      throw new RuntimeException("this will never happen")
    }

    override def comment(node: XdmNode): Unit = {
      throw new RuntimeException("this will never happen")
    }

    override def pi(node: XdmNode): Unit = {
      throw new RuntimeException("this will never happen")
    }
  }

  private class XMLViewportItem(item: XdmNode, meta: XProcMetadata) extends ViewportItem {
    private var replacement: XdmNode = _

    def getReplacement: XdmNode = replacement

    override def getItem: Any = {
      item
    }
    override def getMetadata: Metadata = meta
    override def putItems(items: List[Any]): Unit = {
      val lb = ListBuffer.empty[XdmNode]
      for (item <- items) {
        item match {
          case node: XdmNode =>
            lb += node
          case _ =>
            throw XProcException.xdBadViewportResult(context.location)
        }
      }

      if (lb.size == 1) {
        replacement = lb.head
      } else {
        val builder = new SaxonTreeBuilder(config)
        if (lb.isEmpty) {
          builder.startDocument(None)
        } else {
          builder.startDocument(lb.head.getBaseURI)
        }
        for (item <- lb) {
          builder.addSubtree(item)
        }
        builder.endDocument()
        replacement = builder.result
      }
    }
  }
}

File: xmlcalabash/src/main/kotlin/com/xmlcalabash/util/XmlViewportComposer.kt
package com.xmlcalabash.util

import com.xmlcalabash.config.XProcStepConfiguration
import com.xmlcalabash.datamodel.MediaType
import com.xmlcalabash.documents.DocumentProperties
import com.xmlcalabash.documents.XProcDocument
import com.xmlcalabash.exceptions.XProcError
import com.xmlcalabash.namespace.Ns
import com.xmlcalabash.namespace.NsCx
import com.xmlcalabash.runtime.ProcessMatch
import com.xmlcalabash.runtime.ProcessMatchingNodes
import net.sf.saxon.om.AttributeMap
import net.sf.saxon.s9api.QName
import net.sf.saxon.s9api.XdmNode
import net.sf.saxon.s9api.XdmValue

class XmlViewportComposer(val stepConfig: XProcStepConfiguration, val match: String, val bindings: Map<QName,XdmValue>) {
    var itemIndex = 0

    private var _matcher: ProcessMatch? = null
    private var matcher: ProcessMatch
        get() = _matcher ?: throw RuntimeException("Configuration error: no matcher defined")
        set(value) {
            _matcher = value
        }

    var decomposed: XdmNode? = null
    lateinit var properties: DocumentProperties
    lateinit var contentType: MediaType
    val viewportItems = mutableListOf<XmlViewportItem>()

    fun decompose(document: XProcDocument): List<XmlViewportItem> {
        if (document.value !is XdmNode) {
            throw XProcError.xdViewportNotXml().exception()
        }

        contentType = document.contentType ?: MediaType.XML
        if (!contentType.htmlContentType() && !contentType.xmlContentType()) {
            throw XProcError.xdViewportNotXml().exception()
        }

        properties = document.properties

        val node = document.value as XdmNode
        matcher = ProcessMatch(stepConfig, Decomposer(), stepConfig.inscopeNamespaces, bindings)
        matcher.process(node, match)
        decomposed = matcher.result
        return viewportItems
    }

    fun recompose(): XProcDocument {
        matcher = ProcessMatch(stepConfig, Recomposer(), mapOf("cx" to NsCx.namespace), bindings)
        matcher.process(decomposed!!, "cx:viewport")

        // The result is text iff the match was the document node and the result is text
        if (match == "/") {
            if (S9Api.isTextDocument(matcher.result)) {
                if (properties.has(Ns.serialization)) {
                    properties.remove(Ns.serialization)
                }
                return XProcDocument.ofText(matcher.result, stepConfig, MediaType.TEXT, properties)
            } else {
                return XProcDocument.ofXml(matcher.result, stepConfig, contentType, properties)
            }
        }
        return XProcDocument.ofXml(matcher.result, stepConfig, contentType, properties)
    }

    inner class Decomposer(): ProcessMatchingNodes {
        private fun insertMarker() {
            matcher.addStartElement(NsCx.viewport, stepConfig.attributeMap(mapOf(Ns.index to itemIndex.toString())))
            itemIndex++
        }

        override fun startDocument(node: XdmNode): Boolean {
            viewportItems.add(XmlViewportItem(stepConfig, node))
            insertMarker()
            matcher.addEndElement()
            return false
        }

        override fun endDocument(node: XdmNode) {
            matcher.endDocument()
        }

        override fun startElement(node: XdmNode, attributes: AttributeMap): Boolean {
            val builder = SaxonTreeBuilder(stepConfig)
            builder.startDocument(node.baseURI)
            builder.addSubtree(node)
            builder.endDocument()
            viewportItems.add(XmlViewportItem(stepConfig, builder.result))
            insertMarker()
            return false
        }

        override fun attributes(
            node: XdmNode,
            matchingAttributes: AttributeMap,
            nonMatchingAttributes: AttributeMap
        ): AttributeMap? {
            throw XProcError.xdViewportOnAttribute(match).exception()
        }

        override fun endElement(node: XdmNode) {
            matcher.addEndElement()
        }

        override fun text(node: XdmNode) {
            viewportItems.add(XmlViewportItem(stepConfig, node))
            insertMarker()
            matcher.addEndElement()
        }

        override fun comment(node: XdmNode) {
            viewportItems.add(XmlViewportItem(stepConfig, node))
            insertMarker()
            matcher.addEndElement()
        }

        override fun pi(node: XdmNode) {
            viewportItems.add(XmlViewportItem(stepConfig, node))
            insertMarker()
            matcher.addEndElement()
        }
    }

    inner class Recomposer: ProcessMatchingNodes {
        private fun processMarker(marker: XdmNode) {
            val index = marker.getAttributeValue(Ns.index)!!.toInt()
            matcher.addSubtree(viewportItems[index].replacement)
        }

        override fun startDocument(node: XdmNode): Boolean {
            matcher.startDocument(node.baseURI)
            return true
        }

        override fun endDocument(node: XdmNode) {
            matcher.endDocument()
        }

        override fun startElement(node: XdmNode, attributes: AttributeMap): Boolean {
            processMarker(node)
            return false
        }

        override fun attributes(
            node: XdmNode,
            matchingAttributes: AttributeMap,
            nonMatchingAttributes: AttributeMap
        ): AttributeMap? {
            throw RuntimeException("Configuration error: recomposer matched attribute")
        }

        override fun endElement(node: XdmNode) {
            // nop
        }

        override fun text(node: XdmNode) {
            throw RuntimeException("Configuration error: recomposer matched text")
        }

        override fun comment(node: XdmNode) {
            throw RuntimeException("Configuration error: recomposer matched comment")
        }

        override fun pi(node: XdmNode) {
            throw RuntimeException("Configuration error: recomposer matched pi")
        }
    }
}